### DISCUSSION TITLE
How RedissonSessionManager with AFTER_REQUEST policy prevents overwrites?

### DISCUSSION THREAD

**Discussion by msevcenko** [redisson/redisson#5176] [Q&A] [answered]
I'd like to ask a couple of related questions to the AFTER_REQUEST upate policy.

RedissonSessionManager with AFTER_REQUEST update policy (particularly with MEMORY read policy), saves the session to redis _every_ time _any_ request has ended (even though the request may not touch the session at all, e.g. static resource that is served from tomcat). And broadcasts this update to all remaining tomcat instances?

My first concern is, doesn't it generate a lot of redis traffic? If I have several tomcat instances in a cluster, rapidly serving many requests such as static resources, against the same session (same user), load-balancing to all instances, they would wildly rewrite the session in redis and broadcast it back and forth, won't they?

The second concern is, if two requests on different instances, save the same session to redis - simultaneously or out of order, what happens? The later wins?

For instance:

1. tc instance A initiates a request, e.g. static resource, not touching the session at all
2. tc instance B initiates another request, changing the session
3. request processing on instance B finishes, saving the changed session to redis, possibly broadcasting the change to A
4. request processing on A is about to terminate; the change from A may or may not be processed yet? isn't there a chance that A instance simply overwrites redis with the older session state, discarding the change produced by instance B?

**Comment by msevcenko** [comment]
I've just found very similar question in issues:

https://github.com/redisson/redisson/issues/2675

As far as I understood the question describes identical scenario. The issue was dismissed which suggest that redis session manager does not attempt to address this issue.

> This case should be handled by developer not Redisson

I am not sure how the application can solve this. It seems to me that the AFTER_REQUEST+MEMORY setup is not very useful in production environments.

**Reply by mrniko** [reply, marked as answer]
https://github.com/redisson/redisson/issues/2675#issuecomment-1638559016

Documentation section describing "updateMode" parameter has been updated

**Reply by msevcenko** [reply]
Thanks for prompt reply. Regarding usefulness of the AFTER_REQUEST+MEMORY mode, I can think of a scenario when it might sort of work:

1. use tomcat cluster with _sticky session_ mechanism; that is, when loadbalancer sends traffic with specific session to particular tomcat instance, it MUST attempt to forward subsequent request to the same instance, as long as the instance is active; only when the instance is detached from the cluster, the traffic fails-over to another instance, which loads last-saved state from redis; also, broadcast on update is pointless, there is no need to keep session on all tomcat instances, if the sticky session mechanism works reliably
2. performance-wise, I think it would be necessary to somehow mark requests that do not touch session (such as static resources) and prevent saving session from those requests; also, processing concurrent requests on the _same_ instance might be subject to the same issue (I didn't find any synchronization mechanism that would prevent that), but it is definitely much easier to implement than in distributed cluster.

**Reply by marcelocastrillo** [reply]
Hi! @msevcenko if I use AFTER_REQUEST+MEMORY + broadcast on update, every tomcat instance would have every session, is that right? That's not  very good for performance. 
When using sticky sessions, I think you may need to broadcast because what happens if the node had a temporary failure and the user is forwarded to another server, makes changes in the session and then goes back to the original server? In that case the session of the first server would not be up to date and wouldn't be loaded from redis.

**Reply by msevcenko** [reply]
> Hi! @msevcenko if I use AFTER_REQUEST+MEMORY + broadcast on update, every tomcat instance would have every session, is that right? That's not very good for performance.

I would think so. It sort of suppresses the point of splitting the load for scalability - you split the traffic on per-user basis, but then you join it again, so every request eventually lands on every instance, at least in terms of session update.

>  I think you may need to broadcast because what happens if the node had a temporary failure and the user is forwarded to another server, makes changes in the session and then goes back to the original server? In that case the session of the first server would not be up to date and wouldn't be loaded from redis.

This is true. The strategy I suggested does not support this scenario. You may consider to tolerate the loss of information during this double-failover, which is not very likely.

Or you may choose not to support "temporary failures". If the load-balancer fails-over from a faulty instance, it should kill it, or at least erase its sessions, or invalidate them somehow before the faulty instance is re-attached to the cluster.

It may still be lesser evil than keeping all sessions up-to-date on all instances.

**Reply by marcelocastrillo** [reply]
@msevcenko thanks for your response.
In terms of performance, does AFTER_REQUEST is worse than DEFAULT updateMode ? 
Also I guess readMode MEMORY is better for performance than REDIS, is that right?

**Reply by msevcenko** [reply]
> In terms of performance, does AFTER_REQUEST is worse than DEFAULT updateMode ?Also I guess readMode MEMORY is better for performance than REDIS, is that right?

I am not affiliated with this project, just a plain user :)

This is what I have found from the doc and implementation:

1) the MEMORY mode keeps sessions in redis, tomcat instance serves as write-through cache. So this may be much more efficient, if you access the session much more often than you modify it; but then you need to broadcast changes to all instances (or you need to use sticky session and make sure that two instances do not have the same session simultaneously, as discussed above). So it is a tradeoff, and MEMORY might be faster in some cases.

2) in case of AFTER_REQUEST, I really dont know; I think that it is generally good idea to defer storing attributes after request, consolidated to single command, even in the MEMORY mode.

Which is what the library _doesn't_ do. The two options are not really orthogonal, the combination AFTER_REQUEST+MEMORY does quite different thing - it saves the whole session, no matter what attributes were changed, if any.

So what I would prefer, is
- to use instance as write through cache, as in MEMORY
- to write attributes changed by setAttribute call, in single command, but AFTER_REQUEST

I guess this combination is not implemented.

Best approximation might be MEMORY+DEFAULT, if you want to optimize for session read.

**Comment by msevcenko** [comment]
Btw, isn't this a typo?

https://github.com/redisson/redisson/blob/3f1fb96e365e703dc4854431388f96d30f97961f/redisson-tomcat/redisson-tomcat-9/src/test/java/org/redisson/tomcat/RedissonSessionManagerTest.java#L56

```
            read(8080, executor, "test", "from_server2");
            read(8080, executor, "test", "from_server2");
```

didn't you want to check the value from both tomcat intances?

**Reply by mrniko** [reply]
Thank you for pointing out. It's fixed now.
